spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java

```diff
@@ -407,6 +428,216 @@ public void close() throws IOException {
     }
   }
 
+  private static class UnpartitionedDeltaWriter extends BaseDeltaWriter {
+    private final PositionDeltaWriter<InternalRow> delegate;
+    private final FileIO io;
+    private final PartitionSpec dataSpec;
+    private final Map<Integer, PartitionSpec> specs;
```

Comment:
Why have `dataSpec` and `specs` for unpartitioned tables? Is this that this may be reading previously partitioned data?

Reply:
Correct, we need `specs` for delete projections and we need `dataSpec` to attach correct spec ID to the produced data file (since we did not reserve a particular spec ID for unpartitioned specs).